rfc-117-containerised-runtime.md

```diff
@@ -0,0 +1,69 @@
+# RFC 000: Use a containerised runtime
+
+## Summary
+
+Use a containerised runtime for hosting GOV.UK applications. GOV.UK applications currently run on virtual machines running Ubuntu Trusty Tahr. There is currently an extended support contract for Ubuntu Trusty which will end in April 2022. Rather than upgrading Ubuntu on the virtual machines GOV.UK should change the application runtime to a containerised platform.
+
+## Problem
+
+GOV.UK currently hosts the majority of its applications and services using virtual machines running Ubuntu 14.04 LTS (Trusty Tahr).
+
+Trusty is out of date, it was released in April 2014 and standard support ended in April 2019. At this time GOV.UK signed up for [Extended Security Maintenance](https://ubuntu.com/esm) (ESM) which provides continued security fixes for high and critical common vulnerabilities and exposures (CVEs) for the [most commonly used packages](https://wiki.ubuntu.com/SecurityTeam/ESM/14.04#A14.04_Infrastructure_ESM_Packages). ESM will end in April 2022.
+
+Running an old operating system is preventing GOV.UK from upgrading key components of our technology stack (MongoDB, Puppet, Python, ...).
+
+Upgrading Ubuntu in the existing infrastructure is difficult for a number of reasons, including but not limited to:
+
+- Multiple applications share the operating system of the same virtual machine. This means it’s not easy to upgrade a single application at a time. Furthermore, the way our [Puppet code](http://github.com/alphagov/govuk-puppet/) is written makes supporting virtual machines on different operating systems at the same time hard to support.
+- GOV.UK currently uses upstart as our init system, but newer versions of Ubuntu switch to systemd. This means GOV.UK will need to do a significant rewrite of our init scripts as part of an upgrade.
+- Our current version of Puppet is 3.8, which is no longer supported (the current version is 6.13). Support for newer operating system versions is likely to be lacking in unsupported versions of Puppet.
+- Other infrastructure as code tools in use (e.g. [Fabric](https://github.com/alphagov/fabric-scripts)) are also currently using very old versions which are very likely to have problems with newer operating systems.
```

Comment:
There are ongoing questions about the need for Fabric scripts and how many are regularly in use - @thomasleese knows more, but Platform Health looked at upgrading them to Python 3 from Python 2 is basically a complete rewrite.

Reply:
I know that upgrading verify-fabric was a major blocker for the Verify operating system upgrade, which is why this is called out explicitly. If we can already get rid of it (before the move to containers) that makes this point less compelling. Nevertheless, with the world as it is now, it's a thing that would make an OS upgrade more difficult.

Reply:
If we were to move to a containerised system the need for an elaborate orchestration tool would be mostly removed.

My feeling is that we have been using fabric to prop up our deployment method because it allows us to created complex orchestration procedures.  

Containerisation tech either manages-out or changes deployment strategy which means the complex orchestration rules/conditionals we created in fabric aren't needed.